### ISSUE TITLE
Upload feature for new Screenshot application

### ISSUE THREAD

**Issue by EbonJaeger** [BuddiesOfBudgie/budgie-desktop#281]
### Budgie version

10.7

### Use-cases

Taking a screenshot and having a button to upload it to an image sharing service (say, Imgur) is very handy; it saves the person taking the screenshot from having to do that work manually. With the recent archival of budgie-screenshot-applet, which had this functionality, it would be great if the built-in screenshot solution could do that, too.

### Proposal

After taking a screenshot:

1. Add a button to the dialog header to upload the image
2. On click, upload the image data to an online service
2a. It may be desirable to have a way to choose a provider, or for the user to set one themselves
3. Read the response from the provider, and either show a link to the image or show an error

### References

The best reference for this is probably [budgie-screenshot-applet](https://github.com/cybre/budgie-screenshot-applet).

**Comment by Solarunit** [comment]
>2a. It may be desirable to have a way to choose a provider, or for the user to set one themselves

Agreed
____

Also would be nice to have:

- several image providers in case one of them unavailable
- a button to copy url of uploaded image
- setting to automatically copy url after upload
- history of the last ~5 screenshots

**Comment by JoshStrobl** [comment]
> several image providers in case one of them unavailable

This is something that was looked into and the problems are:

1. Imgur: Associated with the account in question.
2. Imgbb: Associated with the account, account owner can see all images uploaded by users that use the API key, which is a significant privacy issue.

After auditing a bunch of them, 0x0 came out the best bet as we not only don't need an API key, there is negative impact to user privacy on our part (having the ability to see the uploads of others).  IIRC @joebonrichie has 0x0 uploading working.

The others are reasonable as well, I believe there is a general understanding amongst the team that it should have as much feature parity as possible without there being a trade-off for privacy.

**Comment by joebonrichie** [comment]
Current situation:
I have this implemented with imgur, imagebin and 0x0 as upload providers. Imgur is working using my private api key atm, Imagebin and 0x0 don't need an API key. The current work is biased on Cybre's pre-existing applet however, it has been ported to use libsoup3 rather than libsoup2.

Current blockers:
- I need to make the upload methods async as imagebin was down the other day and hung the entire application until Soup timed out. A progress bar might also be desirable.

Imgur worries:
- There is an daily limit of ~1250 uploads a day, this could be exceeded if we allow all downstreams to use our API key.
- Downstreams could use an API key that isn't created with anonymous usage (https://i.imgur.com/Ied42En.png), meaning it may be possible to snoop on images uploaded using a callback (I haven't tested this yet).

I'll happily share my WIP with anyone who is more motivated than me :)

https://i.imgur.com/rz4dVVY.png

**Comment by EbonJaeger** [comment]
I think it might be best for now to just include providers that don't require API keys. Further down the line, it would be nice to have a way for individuals to put in their own providers with their own keys, but that needs other conversations to happen first.

**Comment by DrunkenAlcoholic** [comment]
Can this also be used as an applet?, I found the original budgie screen shot applet super useful with it in a clicks reach on the panel, I am sure you guys have your reasons for making yet another screenshot tool, when gnome screenshot and budgie screenshot applet are available for budgie, even if you do not make it an applet maybe you can have a panel icons and a setting to auto start minimised to panel on startup

**Comment by fossfreedom** [comment]
You could just drag the budgie screenshot icon in the menu into the icon-task-list if thats what you are using - or plank etc - then its an easy once click accessible.

**Comment by DrunkenAlcoholic** [comment]
lol yeap your right, I didn't think about that

**Comment by Solarunit** [comment]
> I found the original budgie screen shot applet super useful with it in a clicks reach on the panel

Yep, you could make screenshot just with just 2 clicks. It was my all time favorite applet!

**Comment by JoshStrobl** [comment]
Budgie 10 series now in maintenance mode. This isn't happening but we should keep it in mind for 11.
